quests/rl/trainer/trainer.py

```diff
@@ -0,0 +1,188 @@
+    episodes set by the user.
+
+    Args:
+      args (dict): The arguments from the command line parsed by
+        _parse_arguments.
+    """
+    with tf.Session() as sess:
+        env = gym.make(args.game)
+        space_size = env.observation_space.shape
+        action_size = env.action_space.n
+        agent = model.Agent(
+            space_size, action_size, args.learning_rate, args.hidden_neurons,
+            args.gamma, args.explore_decay, args.memory_batch_size,
+            args.memory_size, args.training_mode, sess, args.model_path)
+
+        # Initialize the variables
+        sess.run(tf.global_variables_initializer())
+        saver = tf.train.Saver()
+
+        for episode in range(args.episodes):
+            _play(
+                agent, env, episode, args.render_rate, args.score_print_rate,
+                args.record_rate)
+
+        # Save the model
+        save_path = saver.save(sess, args.model_path)
+
+
+def _play(agent, env, episode, render_rate, score_print_rate, record_rate):
+    """Plays through one episode of the game.
+
+    Initializes TensorFlow, the training agent, and the game environment.
+    The agent plays the game from the starting state for a number of
+    episodes set by the user.
+
+    Args:
+      agent: The actor learning to play in the given environment.
+      env: The environment for the agent to act in. This code is intended for
+        use with OpenAI Gym, but the user can alter the code to provide their
+        own environment.
+      episode (int): The current number of simulaions the agent has completed.
+      render_rate (int): The episode frequency in which to render the game
+        being played by the agent.
+      score_print_rate (int): The episode frequency in which to print the total
+        episode reward.
+      record_rate (int): The episode frequency in which to record the episode
+        reward into tensorboard.
+    """
+    episode_reward = 0  # The total reward for an episode.
+    state = env.reset()  # Set up Environment and get start state.
+    done = False
+
+    while not done:
+        action = agent.act(state)
+        state_prime, reward, done, info = env.step(action)
+        episode_reward += reward
+        agent.memory.add((state, action, reward, state_prime, done))
+
+        # Render only enough to "check in" on the agent to speed up training.
+        if render_rate and episode % render_rate == 0:
+            env.render()
+
+        if done:
+            if score_print_rate and episode % score_print_rate == 0:
+                print(
+                    'Training: {}'.format(agent.train),
+                    'Episode: {}'.format(episode),
+                    'Total reward: {}'.format(episode_reward),
+                )
+        else:
+            state = state_prime  # st+1 is now our current state.
+
+    if agent.train:
+        tensorboard_write = episode % record_rate == 0
```

Comment:
Can you clarify what's happening here? So during play runs, we aren't recording  metrics for tensorboard?

Reply:
This is the math to figure out if the this is a training episode where the reward should be recorded to tensorboard.
I still need to write the load model / eval functions. Come to think of it, it would probably be smarter to make a function to stop learning every x step and do an evaluation run for better. I've added a TODO so I don't forget :)

Reply:
Cool, thanks